core/src/main/java/org/pentaho/di/core/xml/XMLHandler.java

```diff
@@ -0,0 +1,1263 @@
+   */
+  public static byte[] stringToBinary( String string ) throws KettleException {
+    try {
+      byte[] bytes;
+      if ( string == null ) {
+        bytes = new byte[] {};
+      } else {
+        bytes = Base64.decodeBase64( string.getBytes() );
+      }
+      if ( bytes.length > 0 ) {
+        ByteArrayInputStream bais = new ByteArrayInputStream( bytes );
+        GZIPInputStream gzip = new GZIPInputStream( bais );
+        BufferedInputStream bi = new BufferedInputStream( gzip );
+        byte[] result = new byte[] {};
+
+        byte[] extra = new byte[ 1000000 ];
+        int nrExtra = bi.read( extra );
+        while ( nrExtra >= 0 ) {
+          // add it to bytes...
+          //
+          int newSize = result.length + nrExtra;
+          byte[] tmp = new byte[ newSize ];
+          for ( int i = 0; i < result.length; i++ ) {
+            tmp[ i ] = result[ i ];
+          }
+          for ( int i = 0; i < nrExtra; i++ ) {
+            tmp[ result.length + i ] = extra[ i ];
+          }
+
+          // change the result
+          result = tmp;
+          nrExtra = bi.read( extra );
+        }
+        bytes = result;
+        gzip.close();
+      }
+
+      return bytes;
+    } catch ( Exception e ) {
+      throw new KettleException( "Error converting string to binary", e );
+    }
+  }
+
+  public static String buildCDATA( String string ) {
+    return buildCDATA( new StringBuilder(), string ).toString();
+  }
+
+  public static StringBuilder buildCDATA( StringBuilder builder, String string ) {
+    return builder.append( "<![CDATA[" ).append( Const.NVL( string, "" ) ).append( "]]>" );
+  }
+
+  public static String openTag( String tag ) {
+    return openTag( new StringBuilder(), tag ).toString();
+  }
+
+  public static StringBuilder openTag( StringBuilder builder, String tag ) {
+    return builder.append( '<' ).append( tag ).append( '>' );
+  }
+
+  public static StringBuilder openTag( StringBuilder builder, String tag, Map<String, String> attributes ) {
+    if ( attributes == null || attributes.isEmpty() ) {
+      return openTag( builder, tag );
+    }
+    builder.append( '<' ).append( tag );
+    attributes.entrySet().stream()
+      .sorted( Map.Entry.comparingByKey() )
+      .forEach( e -> builder
+        .append( " " )
+        .append( e.getKey() )
+        .append( "=\"" )
+        .append( e.getValue() )
+        .append( "\"" ) );
+    return builder.append( '>' );
+  }
+
+  public static String closeTag( String tag ) {
+    return closeTag( new StringBuilder(), tag ).toString();
+  }
+
+  public static StringBuilder closeTag( StringBuilder builder, String tag ) {
+    return builder.append( "</" ).append( tag ).append( '>' );
+  }
+
+  public static String formatNode( Node node ) throws KettleXMLException {
+    StringWriter sw = new StringWriter();
+    try {
+
+      Transformer t = XMLParserFactoryProducer.createSecureTransformerFactory().newTransformer();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Using TransformerFactory without appropriate configuration can lead to XXE